<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>archR: A package for de novo discovery of different sequence architectures 
composed of combination(s) of sequence elements (motifs). — archR • archR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="archR: A package for de novo discovery of different sequence architectures 
composed of combination(s) of sequence elements (motifs). — archR" />
<meta property="og:description" content="Given a set of DNA sequences, archR enables discovery of
clusters of sequence with characteristic sequence architectures.
Call this function to process a data set using archR." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">archR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/archR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/index.html">Articles</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/snikumbh/archR/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>archR: A package for de novo discovery of different sequence architectures 
composed of combination(s) of sequence elements (motifs).</h1>
    <small class="dont-index">Source: <a href='https://github.com/snikumbh/archR/blob/master/R/archR.R'><code>R/archR.R</code></a>, <a href='https://github.com/snikumbh/archR/blob/master/R/archR_main.R'><code>R/archR_main.R</code></a></small>
    <div class="hidden name"><code>archR.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Given a set of DNA sequences, <code>archR</code> enables discovery of
clusters of sequence with characteristic sequence architectures.</p>
<p>Call this function to process a data set using archR.</p>
    </div>

    <pre class="usage"><span class='fu'>archR</span><span class='op'>(</span>
  <span class='va'>config</span>,
  <span class='va'>seqsMat</span>,
  <span class='va'>seqsRaw</span>,
  seqsPositions <span class='op'>=</span> <span class='cn'>NULL</span>,
  thresholdItr <span class='op'>=</span> <span class='fl'>3</span>,
  setParsimony <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='cn'>FALSE</span>, <span class='cn'>FALSE</span>, <span class='cn'>FALSE</span><span class='op'>)</span>,
  setOCollation <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='cn'>TRUE</span>, <span class='cn'>FALSE</span>, <span class='cn'>FALSE</span><span class='op'>)</span>,
  fresh <span class='op'>=</span> <span class='cn'>TRUE</span>,
  UseOC <span class='op'>=</span> <span class='cn'>NULL</span>,
  oDir <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>config</th>
      <td><p>archR configuration object as returned by
<code><a href='archRSetConfig.html'>archRSetConfig</a></code>. This is a required argument.</p></td>
    </tr>
    <tr>
      <th>seqsMat</th>
      <td><p>A matrix of one-hot encoded sequences with sequences along
columns. This is a required argument.</p></td>
    </tr>
    <tr>
      <th>seqsRaw</th>
      <td><p>A Biostrings::DNAStringSet object. The FASTA sequences as a
DNAStringSet object. This argument required argument.</p></td>
    </tr>
    <tr>
      <th>seqsPositions</th>
      <td><p>Vector. Specify the tick labels for sequence positions.
Default is NULL.</p></td>
    </tr>
    <tr>
      <th>thresholdItr</th>
      <td><p>Numeric. Specify the number of iterations to perform.
Default is 3.</p></td>
    </tr>
    <tr>
      <th>setParsimony</th>
      <td><p>Logical vector. Specify if model selection by
cross-validation should prefer parsimonious solutions. Not required when
stability-based model selection is chosen. Length of the vector should match
number of threshold iterations specified in thresholdItr argument. TRUE
denotes parsimony is followed, FALSE otherwise.</p></td>
    </tr>
    <tr>
      <th>setOCollation</th>
      <td><p>Logical vector. Specify for every iteration of archR
if collation of clusters from outer chunks should be performed. TRUE denotes
clusters are collated, FALSE otherwise.</p></td>
    </tr>
    <tr>
      <th>fresh</th>
      <td><p>Logical. Specify if this is (not) a fresh run. Because
archR enables checkpointing, it is possible to perform additional iterations
upon clusters from an existing archR result (or a checkpoint) object. 
See UseOC argument. For example, when processing a set of FASTA sequences, 
if an earlier call to archR performed two iterations, and now you wish to 
perform a third, the arguments `fresh` and `UseOC` can be used. Simply set 
`fresh` to FALSE and assign the sequence clusters from iteration two from 
the earlier result to `UseOC`. As of v0.1.3, with this setting, archR 
returns a new result object as if the additional iteration performed is the 
only iteration.</p></td>
    </tr>
    <tr>
      <th>UseOC</th>
      <td><p>List. Clusters to be further processed with archR. These can 
be from a previous archR result (in which case use 
archRresult$clustSol$clusters), or simply clusters from any other method.
Warning: This has not been rigorously tested yet (v0.1.3).</p></td>
    </tr>
    <tr>
      <th>oDir</th>
      <td><p>Character. Specify the output directory with its path. archR
will create this directory. If a directory with the given name exists at the
given location, archR will add a suffix to the directory name. This
change is reported to the user. Default is NULL. When NULL, just the result
is returned, and no plots or checkpoints or result is written to disk.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A nested list of elements as follows:</p><dl>
<dt>seqsClustLabels</dt><dd><p>A list with cluster labels for all sequences per 
iteration of archR. The cluster labels as stored as characters.</p></dd>

<dt>clustBasisVectors</dt><dd><p>A list with information on NMF basis vectors per 
iteration of archR. Per iteration, there are two variables `nBasisVectors` 
storing the number of basis vectors after model selection,
and `basisVectors`, a matrix storing the basis vectors themselves. 
Dimensions of the `basisVectors` matrix are 4*L x nBasisVectors 
(mononucleotide case) or 16*L x nBasisVectors (dinucleotide case).</p></dd>
 
<dt>clustSol</dt><dd><p>The clustering solution obtained upon processing the raw 
clusters from the last iteration of archR's result. This is handled 
internally by the function <code><a href='reorder_archRresult.html'>reorder_archRresult</a></code> using the
default setting of Euclidean distance and ward.D linkage hierarchical 
clustering.</p></dd>
 
<dt>rawSeqs</dt><dd><p>The input sequences as a DNAStringSet object.</p></dd>

<dt>timeInfo</dt><dd><p>Stores the time taken (in minutes) for processing each 
iteration. This element is added only if `timeFlag` is set to TRUE in 
config.</p></dd>

<dt>config</dt><dd><p>The configuration used for processing.</p></dd>
<dt>call</dt><dd><p>The function call itself.</p></dd>

</dl>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The archR package provides three categories of important functions:
related to data preparation and manipulation, performing non-negative matrix
factorization, performing clustering, and visualization-related functions.</p>
    <h2 class="hasAnchor" id="functions-for-data-preparation-and-manipulation"><a class="anchor" href="#functions-for-data-preparation-and-manipulation"></a>Functions for data preparation and manipulation</h2>

    

<ul>
<li><p><code><a href='prepare_data_from_FASTA.html'>prepare_data_from_FASTA</a></code></p></li>
<li><p><code><a href='get_one_hot_encoded_seqs.html'>get_one_hot_encoded_seqs</a></code></p></li>
</ul>

    <h2 class="hasAnchor" id="functions-for-visualizations"><a class="anchor" href="#functions-for-visualizations"></a>Functions for visualizations</h2>

    

<ul>
<li><p><code><a href='plot_ggseqlogo.html'>plot_ggseqlogo</a></code></p></li>
<li><p><code><a href='plot_ggheatmap.html'>plot_ggheatmap</a></code></p></li>
<li><p><code><a href='viz_basis_vectors_as_heatmap.html'>viz_basis_vectors_as_heatmap</a></code></p></li>
<li><p><code><a href='viz_basis_vectors_as_seqlogo.html'>viz_basis_vectors_as_seqlogo</a></code></p></li>
<li><p><code><a href='viz_basis_vectors_in_combined_heatmaps_seqlogos.html'>viz_basis_vectors_in_combined_heatmaps_seqlogos</a></code></p></li>
<li><p><code><a href='plot_arch_for_clusters.html'>plot_arch_for_clusters</a></code></p></li>
<li><p><code><a href='viz_seqs_as_acgt_mat_from_seqs.html'>viz_seqs_as_acgt_mat_from_seqs</a></code></p></li>
</ul>


    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='va'>inputFastaFilename</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>, <span class='st'>"example_data.fa"</span>, 
                        package <span class='op'>=</span> <span class='st'>"archR"</span>, mustWork <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>

<span class='co'># Specifying 'dinuc' generates dinucleotide features</span>
<span class='va'>inputSeqsMat</span> <span class='op'>&lt;-</span> <span class='fu'>archR</span><span class='fu'>::</span><span class='fu'><a href='prepare_data_from_FASTA.html'>prepare_data_from_FASTA</a></span><span class='op'>(</span><span class='va'>inputFastaFilename</span>,
    sinuc_or_dinuc <span class='op'>=</span> <span class='st'>"dinuc"</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Sequences OK, </span></div><div class='output co'>#&gt; <span class='message'>Read 200 sequences</span></div><div class='output co'>#&gt; <span class='message'>Generating dinucleotide profiles</span></div><div class='input'>
<span class='va'>inputSeqsRaw</span> <span class='op'>&lt;-</span> <span class='fu'>archR</span><span class='fu'>::</span><span class='fu'><a href='prepare_data_from_FASTA.html'>prepare_data_from_FASTA</a></span><span class='op'>(</span><span class='va'>inputFastaFilename</span>, 
    rawSeq <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>

<span class='co'># Set archR configuration</span>
<span class='va'>archRconfig</span> <span class='op'>&lt;-</span> <span class='fu'>archR</span><span class='fu'>::</span><span class='fu'><a href='archRSetConfig.html'>archRSetConfig</a></span><span class='op'>(</span>
    parallelize <span class='op'>=</span> <span class='cn'>TRUE</span>,
    nCoresUse <span class='op'>=</span> <span class='fl'>2</span>,
    nIterationsUse <span class='op'>=</span> <span class='fl'>100</span>,
    kMin <span class='op'>=</span> <span class='fl'>1</span>,
    kMax <span class='op'>=</span> <span class='fl'>20</span>,
    modSelType <span class='op'>=</span> <span class='st'>"stability"</span>,
    tol <span class='op'>=</span> <span class='fl'>10</span><span class='op'>^</span><span class='op'>-</span><span class='fl'>4</span>,
    bound <span class='op'>=</span> <span class='fl'>10</span><span class='op'>^</span><span class='op'>-</span><span class='fl'>8</span>,
    innerChunkSize <span class='op'>=</span> <span class='fl'>100</span>,
    flags <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>debugFlag <span class='op'>=</span> <span class='cn'>TRUE</span>, timeFlag <span class='op'>=</span> <span class='cn'>TRUE</span>, verboseFlag <span class='op'>=</span> <span class='cn'>TRUE</span>,
        plotVerboseFlag <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
<span class='op'>)</span>

<span class='co'># Run archR </span>
<span class='va'>archRresult</span> <span class='op'>&lt;-</span> <span class='fu'>archR</span><span class='fu'>::</span><span class='fu'>archR</span><span class='op'>(</span>config <span class='op'>=</span> <span class='va'>archRconfig</span>,
                          seqsMat <span class='op'>=</span> <span class='va'>inputSeqsMat</span>,
                          seqsRaw <span class='op'>=</span> <span class='va'>inputSeqsRaw</span>,
                          seqsPositions <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq</a></span><span class='op'>(</span><span class='fl'>1</span>,<span class='fl'>100</span>,by<span class='op'>=</span><span class='fl'>1</span><span class='op'>)</span>,
                          thresholdItr <span class='op'>=</span> <span class='fl'>2</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Asked for parallelization</span></div><div class='output co'>#&gt; <span class='message'>Making cluster here</span></div><div class='output co'>#&gt; <span class='message'>Parallelization with 2 cores</span></div><div class='output co'>#&gt; <span class='message'>list(con = 5, host = "localhost", rank = 1)list(con = 6, host = "localhost", rank = 2)</span></div><div class='output co'>#&gt; <span class='message'>Model selection by factor stability</span></div><div class='output co'>#&gt; <span class='message'>Tolerance: 1e-04 &amp; Bound: 1e-08</span></div><div class='output co'>#&gt; <span class='message'>=== Iteration 1, 1 chunk(s) ===</span></div><div class='output co'>#&gt; <span class='message'>[Outer chunk 1/1] [Size: 200]</span></div><div class='output co'>#&gt; <span class='message'>Decision: Processing</span></div><div class='output co'>#&gt; <span class='message'>[Inner chunk 1/2] [Size: 100]</span></div><div class='output co'>#&gt; <span class='message'>Performing stability-based model selection</span></div><div class='output co'>#&gt; <span class='message'>Tolerance: 1e-04 &amp; Bound : 1e-08</span></div><div class='output co'>#&gt; <span class='message'>Checking K = 1</span></div><div class='output co'>#&gt; <span class='message'>AmariTypeDist : 2.15315980533364e-17</span></div><div class='output co'>#&gt; <span class='message'>Checking K = 2</span></div><div class='output co'>#&gt; <span class='message'>AmariTypeDist : 7.48895894792606e-17</span></div><div class='output co'>#&gt; <span class='message'>Change is : 0</span></div><div class='output co'>#&gt; <span class='message'>Checking K = 3</span></div><div class='output co'>#&gt; <span class='message'>AmariTypeDist : 1.82659723485804e-16</span></div><div class='output co'>#&gt; <span class='message'>Change is : 1</span></div><div class='output co'>#&gt; <span class='message'>Checking K = 4</span></div><div class='output co'>#&gt; <span class='message'>AmariTypeDist : 4.8923605148522e-07</span></div><div class='output co'>#&gt; <span class='message'>This amariType Distance = 4.8923605148522e-07</span></div><div class='output co'>#&gt; <span class='message'>Best K for this subset: 3</span></div><div class='output co'>#&gt; <span class='message'>Fetching 3 clusters</span></div><div class='output co'>#&gt; <span class='message'>Best Q2 giving run found: 0.0416069700312701</span></div><div class='output co'>#&gt; <span class='message'>Fetching 3 cluster(s) w/ NMF scores</span></div><div class='output co'>#&gt; <span class='message'>[Inner chunk 2/2] [Size: 100]</span></div><div class='output co'>#&gt; <span class='message'>Performing stability-based model selection</span></div><div class='output co'>#&gt; <span class='message'>Tolerance: 1e-04 &amp; Bound : 1e-08</span></div><div class='output co'>#&gt; <span class='message'>Checking K = 1</span></div><div class='output co'>#&gt; <span class='message'>AmariTypeDist : 2.51201977288924e-17</span></div><div class='output co'>#&gt; <span class='message'>Checking K = 2</span></div><div class='output co'>#&gt; <span class='message'>AmariTypeDist : 7.13234185516767e-18</span></div><div class='output co'>#&gt; <span class='message'>Change is : 1</span></div><div class='output co'>#&gt; <span class='message'>Checking K = 3</span></div><div class='output co'>#&gt; <span class='message'>AmariTypeDist : 2.73885515838114e-14</span></div><div class='output co'>#&gt; <span class='message'>Change is : 4</span></div><div class='output co'>#&gt; <span class='message'>magnitude change higher than tolerance, 4</span></div><div class='output co'>#&gt; <span class='message'>This amariType Distance = 2.73885515838114e-14</span></div><div class='output co'>#&gt; <span class='message'>Would have choosen bestK as : 2</span></div><div class='output co'>#&gt; <span class='message'>Checking K = 4</span></div><div class='output co'>#&gt; <span class='message'>AmariTypeDist : 1.32872604209591e-08</span></div><div class='output co'>#&gt; <span class='message'>This amariType Distance = 1.32872604209591e-08</span></div><div class='output co'>#&gt; <span class='message'>Best K for this subset: 3</span></div><div class='output co'>#&gt; <span class='message'>Fetching 3 clusters</span></div><div class='output co'>#&gt; <span class='message'>Best Q2 giving run found: 0.0496099954482235</span></div><div class='output co'>#&gt; <span class='message'>Fetching 3 cluster(s) w/ NMF scores</span></div><div class='output co'>#&gt; <span class='message'>[Outer chunk 1/1] complete</span></div><div class='output co'>#&gt; <span class='message'>[Iteration 1] complete</span></div><div class='output co'>#&gt; <span class='message'>[Outer chunk 1/1] complete</span></div><div class='output co'>#&gt; <span class='message'>current total factors: 6</span></div><div class='output co'>#&gt; <span class='message'>Current total chunks for next iteration: 6</span></div><div class='output co'>#&gt; <span class='message'>Managing clusters from outer chunk(s)</span></div><div class='output co'>#&gt; <span class='message'>Collation this iter:TRUE; #InnerChunks: 2; #OuterChunks: 1</span></div><div class='output co'>#&gt; <span class='message'>Decision for outer chunk collation: Yes</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'>Hierarchical clustering</span></div><div class='output co'>#&gt; <span class='message'>RE-ORDERING CLUSTERS</span></div><div class='output co'>#&gt; <span class='message'>Cluster order by ward.D linkage w/ euclid distance:</span></div><div class='output co'>#&gt; <span class='message'>New order: 2, 5, 1, 4, 3, 6</span></div><div class='output co'>#&gt; <span class='message'>2, 5, 2.94351671321916</span></div><div class='output co'>#&gt; <span class='message'>5, 1, 5.93222437689053</span></div><div class='output co'>#&gt; <span class='message'>1, 4, 2.4270593689496</span></div><div class='output co'>#&gt; <span class='message'>4, 3, 5.53264304797784</span></div><div class='output co'>#&gt; <span class='message'>3, 6, 3.00071774134856</span></div><div class='output co'>#&gt; <span class='message'>New order: 2, 5, 1, 4, 3, 6</span></div><div class='output co'>#&gt; <span class='message'>=== CALLING get_clusters FUNCTION === </span></div><div class='output co'>#&gt; <span class='message'>Max. silhouette score at index = 13, height:3.0270593689496</span></div><div class='output co'>#&gt; <span class='message'>Final #Clusters using silhouette values: 3</span></div><div class='output co'>#&gt; <span class='message'>c(1, 4)</span>
#&gt; <span class='message'>c(2, 5)</span>
#&gt; <span class='message'>c(3, 6)</span></div><div class='output co'>#&gt; <span class='message'>Final #Clusters: 3</span></div><div class='output co'>#&gt; <span class='message'>c(1, 4)</span>
#&gt; <span class='message'>c(2, 5)</span>
#&gt; <span class='message'>c(3, 6)</span></div><div class='output co'>#&gt; <span class='message'>#Clusters: 3</span></div><div class='output co'>#&gt; <span class='message'>=== DONE ===</span></div><div class='output co'>#&gt; <span class='message'>ClustList: c(1, 4)</span>
#&gt; <span class='message'>c(2, 5)</span>
#&gt; <span class='message'>c(3, 6)</span></div><div class='output co'>#&gt; <span class='message'>=== This is what I got ===</span></div><div class='output co'>#&gt; <span class='message'>c(3, 5, 7, 8, 10, 12, 13, 16, 17, 18, 19, 22, 25, 30, 31, 36, 37, 38, 39, 41, 43, 45, 47, 49, 51, 53, 54, 55, 56, 58, 59, 60, 61, 64, 67, 69, 70, 73, 77, 84, 89, 91, 94, 95, 98)</span>
#&gt; <span class='message'>c(1, 2, 6, 11, 20, 23, 32, 46, 50, 52, 63, 68, 71, 72, 75, 80, 82, 83, 85, 86, 87, 88, 92, 99)</span>
#&gt; <span class='message'>c(4, 9, 14, 15, 21, 24, 26, 27, 28, 29, 33, 34, 35, 40, 42, 44, 48, 57, 62, 65, 66, 74, 76, 78, 79, 81, 90, 93, 96, 97, 100)</span>
#&gt; <span class='message'>c(103, 106, 107, 109, 115, 117, 118, 119, 120, 121, 122, 124, 125, 126, 128, 130, 132, 133, 134, 135, 136, 140, 141, 142, 144, 150, 152, 153, 154, 156, 159, 160, 161, 162, 164, 166, 170, 173, 175, 177, 180, 183, 184, 185, 186, 188, 189, 191, 193, 194, 195, 196, 198, 199, 200)</span>
#&gt; <span class='message'>c(102, 104, 105, 112, 114, 116, 123, 131, 137, 138, 143, 145, 146, 147, 148, 151, 158, 165, 167, 168, 171, 176, 179, 182, 190, 197)</span>
#&gt; <span class='message'>c(101, 108, 110, 111, 113, 127, 129, 139, 149, 155, 157, 163, 169, 172, 174, 178, 181, 187, 192)</span></div><div class='output co'>#&gt; <span class='message'>=== INFO ===</span></div><div class='output co'>#&gt; <span class='message'>1, 4</span></div><div class='output co'>#&gt; <span class='message'>Size:100</span></div><div class='output co'>#&gt; <span class='message'>=== INFO ===</span></div><div class='output co'>#&gt; <span class='message'>2, 5</span></div><div class='output co'>#&gt; <span class='message'>Size:50</span></div><div class='output co'>#&gt; <span class='message'>=== INFO ===</span></div><div class='output co'>#&gt; <span class='message'>3, 6</span></div><div class='output co'>#&gt; <span class='message'>Size:50</span></div><div class='output co'>#&gt; <span class='message'>=== I am returning ===</span></div><div class='output co'>#&gt; <span class='message'>c(3, 5, 7, 8, 10, 12, 13, 16, 17, 18, 19, 22, 25, 30, 31, 36, 37, 38, 39, 41, 43, 45, 47, 49, 51, 53, 54, 55, 56, 58, 59, 60, 61, 64, 67, 69, 70, 73, 77, 84, 89, 91, 94, 95, 98, 103, 106, 107, 109, 115, 117, 118, 119, 120, 121, 122, 124, 125, 126, 128, 130, 132, 133, 134, 135, 136, 140, 141, 142, 144, 150, 152, 153, 154, 156, 159, 160, 161, 162, 164, 166, 170, 173, 175, 177, 180, 183, 184, 185, 186, 188, 189, 191, 193, 194, 195, 196, 198, 199, 200)</span>
#&gt; <span class='message'>c(1, 2, 6, 11, 20, 23, 32, 46, 50, 52, 63, 68, 71, 72, 75, 80, 82, 83, 85, 86, 87, 88, 92, 99, 102, 104, 105, 112, 114, 116, 123, 131, 137, 138, 143, 145, 146, 147, 148, 151, 158, 165, 167, 168, 171, 176, 179, 182, 190, 197)</span>
#&gt; <span class='message'>c(4, 9, 14, 15, 21, 24, 26, 27, 28, 29, 33, 34, 35, 40, 42, 44, 48, 57, 62, 65, 66, 74, 76, 78, 79, 81, 90, 93, 96, 97, 100, 101, 108, 110, 111, 113, 127, 129, 139, 149, 155, 157, 163, 169, 172, 174, 178, 181, 187, 192)</span></div><div class='output co'>#&gt; <span class='message'>Updating sequence cluster labels</span></div><div class='output co'>#&gt; <span class='message'>#Clusters: 3</span></div><div class='output co'>#&gt; <span class='message'>Iteration 1 completed in 1.27 mins</span></div><div class='output co'>#&gt; <span class='message'>1.27 mins ellapsed since start</span></div><div class='output co'>#&gt; <span class='message'>=== Iteration 2, 3 chunk(s) ===</span></div><div class='output co'>#&gt; <span class='message'>[Outer chunk 1/3] [Size: 100]</span></div><div class='output co'>#&gt; <span class='message'>Decision: Processing</span></div><div class='output co'>#&gt; <span class='message'>[Inner chunk 1/1] [Size: 100]</span></div><div class='output co'>#&gt; <span class='message'>Performing stability-based model selection</span></div><div class='output co'>#&gt; <span class='message'>Tolerance: 1e-04 &amp; Bound : 1e-08</span></div><div class='output co'>#&gt; <span class='message'>Checking K = 1</span></div><div class='output co'>#&gt; <span class='message'>AmariTypeDist : 1.61711272879745e-17</span></div><div class='output co'>#&gt; <span class='message'>Checking K = 2</span></div><div class='output co'>#&gt; <span class='message'>AmariTypeDist : 4.75160915776556e-11</span></div><div class='output co'>#&gt; <span class='message'>Change is : 6</span></div><div class='output co'>#&gt; <span class='message'>magnitude change higher than tolerance, 4</span></div><div class='output co'>#&gt; <span class='message'>This amariType Distance = 4.75160915776556e-11</span></div><div class='output co'>#&gt; <span class='message'>Would have choosen bestK as : 1</span></div><div class='output co'>#&gt; <span class='message'>Checking K = 3</span></div><div class='output co'>#&gt; <span class='message'>AmariTypeDist : 0.0035786011862791</span></div><div class='output co'>#&gt; <span class='message'>This amariType Distance = 0.0035786011862791</span></div><div class='output co'>#&gt; <span class='message'>Best K for this subset: 2</span></div><div class='output co'>#&gt; <span class='message'>Fetching 2 clusters</span></div><div class='output co'>#&gt; <span class='message'>Best Q2 giving run found: 0.114541701362898</span></div><div class='output co'>#&gt; <span class='message'>Fetching 2 cluster(s) w/ NMF scores</span></div><div class='output co'>#&gt; <span class='message'>[Outer chunk 1/3] complete</span></div><div class='output co'>#&gt; <span class='message'>[Outer chunk 1/3] complete</span></div><div class='output co'>#&gt; <span class='message'>current total factors: 2</span></div><div class='output co'>#&gt; <span class='message'>Current total chunks for next iteration: 2</span></div><div class='output co'>#&gt; <span class='message'>[Outer chunk 2/3] [Size: 50]</span></div><div class='output co'>#&gt; <span class='message'>Decision: Processing</span></div><div class='output co'>#&gt; <span class='message'>[Inner chunk 1/1] [Size: 50]</span></div><div class='output co'>#&gt; <span class='message'>Performing stability-based model selection</span></div><div class='output co'>#&gt; <span class='message'>Tolerance: 1e-04 &amp; Bound : 1e-08</span></div><div class='output co'>#&gt; <span class='message'>Checking K = 1</span></div><div class='output co'>#&gt; <span class='message'>AmariTypeDist : 1.59019823123078e-17</span></div><div class='output co'>#&gt; <span class='message'>Checking K = 2</span></div><div class='output co'>#&gt; <span class='message'>AmariTypeDist : 2.57849364933938e-06</span></div><div class='output co'>#&gt; <span class='message'>This amariType Distance = 2.57849364933938e-06</span></div><div class='output co'>#&gt; <span class='message'>Best K for this subset: 1</span></div><div class='output co'>#&gt; <span class='message'>Fetching 1 clusters</span></div><div class='output co'>#&gt; <span class='message'>Best Q2 giving run found: 0.197789465153941</span></div><div class='output co'>#&gt; <span class='message'>Fetching 1 cluster(s) w/ NMF scores</span></div><div class='output co'>#&gt; <span class='message'>[Outer chunk 2/3] complete</span></div><div class='output co'>#&gt; <span class='message'>[Outer chunk 2/3] complete</span></div><div class='output co'>#&gt; <span class='message'>current total factors: 3</span></div><div class='output co'>#&gt; <span class='message'>Current total chunks for next iteration: 3</span></div><div class='output co'>#&gt; <span class='message'>[Outer chunk 3/3] [Size: 50]</span></div><div class='output co'>#&gt; <span class='message'>Decision: Processing</span></div><div class='output co'>#&gt; <span class='message'>[Inner chunk 1/1] [Size: 50]</span></div><div class='output co'>#&gt; <span class='message'>Performing stability-based model selection</span></div><div class='output co'>#&gt; <span class='message'>Tolerance: 1e-04 &amp; Bound : 1e-08</span></div><div class='output co'>#&gt; <span class='message'>Checking K = 1</span></div><div class='output co'>#&gt; <span class='message'>AmariTypeDist : 3.17142496327267e-17</span></div><div class='output co'>#&gt; <span class='message'>Checking K = 2</span></div><div class='output co'>#&gt; <span class='message'>AmariTypeDist : 7.67125992969052e-05</span></div><div class='output co'>#&gt; <span class='message'>This amariType Distance = 7.67125992969052e-05</span></div><div class='output co'>#&gt; <span class='message'>Best K for this subset: 1</span></div><div class='output co'>#&gt; <span class='message'>Fetching 1 clusters</span></div><div class='output co'>#&gt; <span class='message'>Best Q2 giving run found: 0.153271954789996</span></div><div class='output co'>#&gt; <span class='message'>Fetching 1 cluster(s) w/ NMF scores</span></div><div class='output co'>#&gt; <span class='message'>[Outer chunk 3/3] complete</span></div><div class='output co'>#&gt; <span class='message'>[Iteration 2] complete</span></div><div class='output co'>#&gt; <span class='message'>[Outer chunk 3/3] complete</span></div><div class='output co'>#&gt; <span class='message'>current total factors: 4</span></div><div class='output co'>#&gt; <span class='message'>Current total chunks for next iteration: 4</span></div><div class='output co'>#&gt; <span class='message'>Managing clusters from outer chunk(s)</span></div><div class='output co'>#&gt; <span class='message'>Collation this iter:FALSE; #InnerChunks: 1; #OuterChunks: 3</span></div><div class='output co'>#&gt; <span class='message'>Decision for outer chunk collation: No</span></div><div class='output co'>#&gt; <span class='message'>Updating sequence cluster labels</span></div><div class='output co'>#&gt; <span class='message'>#Clusters: 4</span></div><div class='output co'>#&gt; <span class='message'>Iteration 2 completed in 0.814 mins</span></div><div class='output co'>#&gt; <span class='message'>2.08 mins ellapsed since start</span></div><div class='output co'>#&gt; <span class='message'>Hierarchical clustering</span></div><div class='output co'>#&gt; <span class='message'>RE-ORDERING CLUSTERS</span></div><div class='output co'>#&gt; <span class='message'>Cluster order by ward.D linkage w/ euclid distance:</span></div><div class='output co'>#&gt; <span class='message'>New order: 3, 4, 1, 2</span></div><div class='output co'>#&gt; <span class='message'>3, 4, 6.00792293609081</span></div><div class='output co'>#&gt; <span class='message'>4, 1, 5.669887508159</span></div><div class='output co'>#&gt; <span class='message'>1, 2, 3.67152457274261</span></div><div class='output co'>#&gt; <span class='message'>New order: 3, 4, 1, 2</span></div><div class='output co'>#&gt; <span class='message'>=== CALLING get_clusters FUNCTION === </span></div><div class='output co'>#&gt; <span class='message'>No element pairs close enough by given distance threshold: 3</span></div><div class='output co'>#&gt; <span class='message'>#Clusters: 4</span></div><div class='output co'>#&gt; <span class='message'>=== DONE ===</span></div><div class='output co'>#&gt; <span class='message'>ClustList: 3</span>
#&gt; <span class='message'>4</span>
#&gt; <span class='message'>1</span>
#&gt; <span class='message'>2</span></div><div class='output co'>#&gt; <span class='message'>Performed factor clustering, returning object:</span></div><div class='output co'>#&gt; <span class='message'>Returning: 3</span>
#&gt; <span class='message'>4</span>
#&gt; <span class='message'>1</span>
#&gt; <span class='message'>2</span></div><div class='output co'>#&gt; <span class='message'>Updating sequence cluster labels</span></div><div class='output co'>#&gt; <span class='message'>#Clusters: 4</span></div><div class='output co'>#&gt; <span class='message'>Stopping cluster...</span></div><div class='output co'>#&gt; <span class='message'>archR exiting, 2.08 mins</span></div><div class='input'>
 
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Sarvesh Nikumbh.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


